aten/src/ATen/Storage.h

```diff
@@ -1,53 +1,122 @@
+//    atomically increment the use count, if it is greater than 0.
+//    If it is not, you must report that the storage is dead.
+//
+
+struct THFinalizer {
+  virtual void operator()() = 0;
+  virtual ~THFinalizer() {};
+};
 
 namespace at {
 
 struct Type;
 
 struct Storage {
-  static const char RESIZABLE = 2;
 
-  Storage() {}
-  Storage(THStorage* storage)
-      : storage(storage) {}
-  Storage(const Storage& other) = delete;
-  virtual ~Storage() {
-    THStorage_free(storage);
+  Storage() = delete;
+  virtual ~Storage() {};
+  Storage(at::Backend, at::ScalarType, ptrdiff_t, at::DataPtr, at::Allocator*, char);
+  Storage(at::Backend, at::ScalarType, ptrdiff_t, at::Allocator*, char);
+  at::ScalarType scalar_type;
+  at::Backend backend_;
+  at::DataPtr data_ptr;
+  ptrdiff_t size;
+  std::atomic<int> refcount;
+  std::atomic<int> weakcount;
+  char flag;
+  at::Allocator* allocator;
+  std::unique_ptr<THFinalizer> finalizer;
+  struct Storage* view;
+  Storage(Storage&) = delete;
+  Storage(const Storage&) = delete;
+  Storage(Storage&&) = delete;
+  Storage(const Storage&&) = delete;
+
+  template <typename T>
+  inline T* data() const {
+    auto scalar_type_T = at::CTypeToScalarType<th::from_type<T>>::to();
+    if (scalar_type != scalar_type_T) {
+      AT_ERROR(
+          "Attempt to access Storage having data type ",
+          at::toString(scalar_type),
+          " as data type ",
+          at::toString(scalar_type_T));
+    }
+    return unsafe_data<T>();
   }
+
+  template <typename T>
+  inline T* unsafe_data() const {
+    return static_cast<T*>(this->data_ptr.get());
+  }
+  static const char REFCOUNTED = 1;
+  static const char RESIZABLE = 2;
+
+//  virtual ~Storage() {
+//    THStorage_free(storage);
+//  }
   void operator=(const Storage&) = delete;
 
-  virtual size_t elementSize() const = 0;
-  size_t size() const {
-    return storage->size;
+  virtual size_t elementSize() const {
+    return at::elementSize(scalar_type);
+  }
+  //TODO: Rename to size() and size to size_
+  size_t get_size() const { 
+    return size;
   };
   void* data() {
-    return storage->data_ptr.get();
+    return data_ptr.get();
   };
   const void* data() const {
-    return storage->data_ptr.get();
+    return data_ptr.get();
   };
-  void* unsafeGetTH(bool retain_) const {
-    if (retain_) {
-      THStorage_retain(storage);
+  void* unsafeGetTH(bool retain_) {
+    if (retain_ && flag & REFCOUNTED) {
+      ++refcount;
     }
-    return storage;
+    return this;
   }
   void retain() {
-    THStorage_retain(storage);
+    if (flag & REFCOUNTED) {
+      ++refcount;
+    }
   }
-  virtual Type & type() const = 0;
+  Type& type() const;
+
   int getDevice() const {
-    return storage->data_ptr.device().index();
+    return data_ptr.device().index();
   }
-  void clear_flag(char flag) {
-    THStorage_clearFlag(storage, flag);
+  void clear_flag(char flag_) {
+    flag &= ~flag_;
```

Comment:
Didn't you have a patch that converted this into a boolean?

Reply:
Might rebase on top of that or the other way around.